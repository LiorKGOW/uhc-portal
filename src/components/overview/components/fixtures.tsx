import React from 'react';

import SomeLogo from '~/styles/images/RedHatOpenShiftGitOpsLogo.svg';

import { DRAWER_PANEL_CONTENT, DrawerPanelContentNode } from './common/DrawerPanelContent';
import PRODUCT_CARD_LOGOS from './common/ProductCardLogos';
import { FEATURED_PRODUCTS_CARDS } from './FeaturedProductsCards/FeaturedProductsCards';

const BASIC_PRODUCT_CARD_PROPS = {
  title: 'example title',
  description: 'example description',
  logo: 'product-overview-card__logo',
  labelText: 'example labelText',
  drawerPanelContent: {
    head: <div data-testid="test-head-div">example drawer panel content head</div>,
    body: <div data-testid="test-body-div">example drawer panel content body</div>,
  },
  openLearnMore: jest.fn(),
  isSelected: false,
  dataTestId: 'some-id',
};

const PRODUCT_CARD_TEST_CASES = {
  UNSELECTED: {
    ...BASIC_PRODUCT_CARD_PROPS,
  },

  SELECTED: {
    ...BASIC_PRODUCT_CARD_PROPS,
    isSelected: true,
  },

  WITHOUT_LABEL_TEXT: {
    ...BASIC_PRODUCT_CARD_PROPS,
    labelText: undefined,
  },
};

const BASIC_DRAWER_PANEL_PROPS = {
  children: <div data-testid="children-of-drawer-panel">some children</div>,
  content: {
    head: <div data-testid="test-head-div">example drawer panel content head</div>,
    body: <div data-testid="test-body-div">example drawer panel content body</div>,
  },
  isOpen: true,
  onClose: jest.fn(),
};

const DRAWER_PANEL_TEST_CASES = {
  OPENED: { ...BASIC_DRAWER_PANEL_PROPS },
};

const BASIC_RECCOMENDED_OPERATORS_CARDS_PROPS = {
  openLearnMore: jest.fn(),
  selectedCardTitle: '',
};

const RECOMMENDED_OPERATORS_CARDS_TEST_CASES = {
  NON_SELECTED: {
    ...BASIC_RECCOMENDED_OPERATORS_CARDS_PROPS,
  },
  GITOPS_SELECTED: {
    ...BASIC_RECCOMENDED_OPERATORS_CARDS_PROPS,
    selectedCardTitle: PRODUCT_CARD_LOGOS.gitops.title,
  },
};

type RecommendedOperatorsCardsNode = {
  title: string;
  description: string;
  logo: string | undefined;
  labelText?: string;
  drawerPanelContent?: DrawerPanelContentNode;
};

const RECOMMENDED_OPERATORS_CARDS_DATA: RecommendedOperatorsCardsNode[] = [
  {
    ...PRODUCT_CARD_LOGOS.gitops,
    description:
      'Integrate git repositories, continuous integration/continuous delivery (CI/CD) tools, and Kubernetes.',
    labelText: 'Free',
    drawerPanelContent: DRAWER_PANEL_CONTENT.gitops,
  },
  {
    ...PRODUCT_CARD_LOGOS.pipelines,
    description:
      'Automate your application delivery using a continuous integration and continuous deployment (CI/CD) framework.',
    labelText: 'Free',
    drawerPanelContent: DRAWER_PANEL_CONTENT.pipelines,
  },
  {
    ...PRODUCT_CARD_LOGOS.serviceMesh,
    description: 'Connect, manage, and observe microservices-based applications in a uniform way.',
    labelText: 'Free',
    drawerPanelContent: DRAWER_PANEL_CONTENT.serviceMesh,
  },
];

const BASIC_FEATURED_PRODUCTS_CARDS_PROPS = {
  openLearnMore: jest.fn(),
  selectedCardTitle: '',
};

const FEATURED_PRODUCTS_CARDS_TEST_CASES = {
  NON_SELECTED: {
    ...BASIC_FEATURED_PRODUCTS_CARDS_PROPS,
  },
  ADVANCED_CLUSTER_SECURITY_SELECTED: {
    ...BASIC_FEATURED_PRODUCTS_CARDS_PROPS,
    selectedCardTitle: PRODUCT_CARD_LOGOS.advancedClusterSecurity.title,
  },
};

const GENERAL_DRAWER_PANEL_HEAD_BASIC = {
  title: 'some title',
  logo: SomeLogo,
};

const GENERAL_DRAWER_PANEL_HEAD_WITH_LINK = {
  ...GENERAL_DRAWER_PANEL_HEAD_BASIC,
  trialButtonLink: 'https://www.redhat.com/en',
};

export {
  PRODUCT_CARD_TEST_CASES,
  DRAWER_PANEL_TEST_CASES,
  RECOMMENDED_OPERATORS_CARDS_TEST_CASES,
  RECOMMENDED_OPERATORS_CARDS_DATA,
  FEATURED_PRODUCTS_CARDS_TEST_CASES,
  FEATURED_PRODUCTS_CARDS,
  GENERAL_DRAWER_PANEL_HEAD_BASIC,
  GENERAL_DRAWER_PANEL_HEAD_WITH_LINK,
};
