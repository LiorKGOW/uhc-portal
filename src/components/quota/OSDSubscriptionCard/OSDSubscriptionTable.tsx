import React from 'react';
import orderBy from 'lodash/orderBy';

import { IRowCell } from '@patternfly/react-table';
import {
  Table as TableDeprecated,
  TableBody as TableBodyDeprecated,
  TableHeader as TableHeaderDeprecated,
} from '@patternfly/react-table/deprecated';

import PopoverHint from '../../common/PopoverHint';

type Props = {
  rows: (React.ReactNode | IRowCell)[][];
  children?: React.ReactNode;
};

const OSDSubscriptionTable = ({ rows, children }: Props) => {
  const sortedRows = orderBy(rows, [0, 1]).map((cells) => ({ cells }));
  const tipText = (
    <>
      <p>Standard: Cluster infrastructure costs paid by Red Hat</p>
      <p>CCS: Cluster infrastructure costs paid by the customer</p>
    </>
  );
  const planType = (
    <>
      Plan type <PopoverHint hint={tipText} />
    </>
  );
  const columns = [
    'Resource type',
    'Resource name',
    'Availability',
    { title: 'Plan type', transforms: [() => ({ children: planType })] },
    'Cluster type',
    { title: 'Used', props: { className: 'quota-table-used' } },
    'Capacity',
  ];

  return (
    <>
      <TableDeprecated aria-label="Quota Table" cells={columns} rows={sortedRows}>
        <TableHeaderDeprecated />
        <TableBodyDeprecated />
      </TableDeprecated>
      {children}
    </>
  );
};

export default OSDSubscriptionTable;
