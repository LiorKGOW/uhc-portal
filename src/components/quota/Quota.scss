.ocm-p-subscriptions {
  .pf-v5-c-card__title {
    font-size: var(--pf-v5-global--FontSize--lg);
    font-weight: var(--pf-v5-global--FontWeight--normal);
  }

  .ocm-l-ocp-subscription__section {
    .osp-list {
      padding-left: var(--pf-v5-global--spacer--2xl);
      li {
        padding-left: var(--pf-v5-global--spacer--sm);
      }
    }
  }

  .ocm-l-osd-subscription__section {
    .content-header {
      font-size: var(--pf-v5-global--FontSize--md);
      font-weight: var(--pf-v5-global--FontWeight--bold);
    }
    .table-container {
      overflow-x: auto;
      .quota-table-used,
      .quota-table-max {
        word-break: normal;
      }
      .warn-icon {
        fill: var(--pf-v5-global--warning-color--100);
      }
    }
  }

  .subscriptions-empty-state {
    background-color: var(--pf-v5-global--BackgroundColor--100);
  }
}
