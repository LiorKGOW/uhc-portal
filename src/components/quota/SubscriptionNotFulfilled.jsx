import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Spinner } from '@redhat-cloud-services/frontend-components';
import {
  EmptyState,
  EmptyStateBody,
  Title,
  Button,
  PageSection,
} from '@patternfly/react-core';
import { overrideErrorMessage, BANNED_USER_CODE } from '../../common/errors';

function SubscriptionNotFulfilled({ data, refresh, marketplace }) {
  const getEmptyState = (title, text, button) => (
    <PageSection className="subscriptions-empty-state">
      <EmptyState>
        <Title headingLevel="h4" size="2xl">
          { title }
        </Title>
        <EmptyStateBody>
          { text }
        </EmptyStateBody>
        { button }
      </EmptyState>
    </PageSection>
  );

  const getErrorText = ({ errorMessage, operationID, internalErrorCode }) => {
    const payload = { code: internalErrorCode };
    const text = BANNED_USER_CODE === payload.code ? overrideErrorMessage(payload) : (
      <>
        <p>
          An error has occured!
          Try again or contact support by including this error message:
        </p>
        <q>
          {errorMessage}
        </q>
        <p>
          {`Operation ID: ${operationID || 'N/A'}`}
        </p>
      </>
    );
    return text;
  };

  const config = {
    account: {
      emptyTitle: 'Unable to retrieve account information',
      errorTitle: 'Unable to retrieve account information',
      text: 'Contact support to verify your account is valid',
    },
    ocp: {
      emptyTitle: 'You do not have any clusters',
      errorTitle: 'Unable to retrieve subscription status',
      text: <p>Create a cluster to get started.</p>,
      emptyButton: <Link to="/install"><Button>Create cluster</Button></Link>,
    },
    osd: {
      emptyTitle: 'You do not have any quota',
      errorTitle: 'Unable to retrieve quota information',
      text: (
        <p>
          <a
            href="https://www.openshift.com/products/dedicated/contact/"
            rel="noreferrer noopener"
            target="_blank"
          >
            Contact sales
          </a>
          {' '}
          to get started with OpenShift Dedicated.
        </p>
      ),
    },
    osdmarketplace: {
      emptyTitle: 'Marketplace on-demand subscriptions not detected',
      errorTitle: 'Unable to retrieve quota information',
      text: (
        <p>
          <p>
            No marketplace subscriptions for OpenShift Dedicated or add-ons
            were found in your account
          </p>
          <br />
          <p>
            <Button component="a" href="https://marketplace.redhat.com/en-us/products/red-hat-openshift-dedicated" variant="primary" rel="noopener noreferrer" target="_blank">
              Enable in Marketplace
            </Button>
          </p>
          <br />
          <a
            href="https://access.redhat.com/documentation/en-us/openshift_cluster_manager/2021-02/html/managing_clusters/assembly-cluster-subscriptions"
            rel="noreferrer noopener"
            target="_blank"
          >
            Learn more
          </a>
        </p>
      ),
    },
  };
  let configType = config[data.type];
  if (marketplace) {
    configType = config[`${data.type}marketplace`];
  }

  let content = null;
  if (data.error) {
    const errorText = getErrorText(data);
    const errorButton = <Button onClick={refresh}>Try again</Button>;
    content = getEmptyState(configType.errorTitle, errorText, errorButton);
  } else if (data.pending) {
    content = <PageSection><Spinner centered /></PageSection>;
  } else if (data.empty) {
    content = getEmptyState(configType.emptyTitle, configType.text, configType.emptyButton);
  }
  return content;
}

SubscriptionNotFulfilled.propTypes = {
  data: PropTypes.object.isRequired,
  refresh: PropTypes.func.isRequired,
  marketplace: PropTypes.bool,
};

export default SubscriptionNotFulfilled;
