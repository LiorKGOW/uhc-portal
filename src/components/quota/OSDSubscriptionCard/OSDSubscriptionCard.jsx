import PropTypes from 'prop-types';
import React, { Component } from 'react';
import get from 'lodash/get';
import startCase from 'lodash/startCase';
import {
  Card,
  CardBody,
  CardTitle,
  Stack,
  StackItem,
} from '@patternfly/react-core';
import {
  ExclamationTriangleIcon,
  ResourcesAlmostEmptyIcon,
  ResourcesAlmostFullIcon,
  ResourcesFullIcon,
  OutlinedCircleIcon,
} from '@patternfly/react-icons';

import SubscriptionNotFulfilled from '../SubscriptionNotFulfilled';
import OSDSubscriptionTable from './OSDSubscriptionTable';
import { billingModels } from '../../../common/subscriptionTypes';

class OSDSubscriptionCard extends Component {
  componentDidMount() {
    this.refresh();
  }

  refresh = () => {
    const { organizationID, fetchQuotaCost } = this.props;
    fetchQuotaCost(organizationID);
  }

  getCapacityIcon = (used, max) => {
    let icon = <ResourcesAlmostEmptyIcon />;
    if (used > max) {
      icon = <ExclamationTriangleIcon className="warn-icon" />;
    } else if (used === max) {
      icon = <ResourcesFullIcon />;
    } else if (used >= max / 2) {
      icon = <ResourcesAlmostFullIcon />;
    } else if (used === 0) {
      icon = <OutlinedCircleIcon />;
    }
    return icon;
  }

  getZoneType = (zoneType) => {
    if (zoneType === 'multi') { return 'multizone'; }
    if (zoneType === 'single') { return 'single zone'; }
    return 'N/A';
  }

  getPlanType = (byoc) => {
    if (byoc === 'rhinfra') { return 'Standard'; }
    if (byoc === 'byoc') { return 'CCS'; }
    if (byoc === 'any') { return 'Any'; }
    return 'N/A';
  }

  render() {
    const { quotaCost, marketplaceQuotaFeature } = this.props;
    let content;
    let rows = [];
    if (quotaCost.fulfilled) {
      rows = quotaCost.items.flatMap((quotaItem) => {
        // filter out quota you neither have nor consume
        if (quotaItem.consumed === 0 && quotaItem.allowed === 0) {
          return [];
        }

        // filter out zero cost related resources
        const relatedResources = get(quotaItem, 'related_resources', []).filter(resource => resource.cost !== 0);
        if (relatedResources.length === 0) {
          return [];
        }

        // filter out marketplace quota unless feature flagged
        if (!marketplaceQuotaFeature) {
          const billingModel = get(relatedResources[0], 'billing_model', billingModels.STANDARD);
          if (billingModel === billingModels.MARKETPLACE) {
            return [];
          }
        }

        // CCS compute.node resource name should show as vCPU
        let resourceName = get(relatedResources[0], 'resource_name');
        if (get(relatedResources[0], 'resource_type') === 'compute.node' && get(relatedResources[0], 'byoc') === 'byoc') {
          resourceName = 'vCPU';
        }

        return [[
          get(relatedResources[0], 'resource_type'),
          resourceName,
          { title: this.getZoneType(get(relatedResources[0], 'availability_zone_type')) },
          this.getPlanType(get(relatedResources[0], 'byoc')),
          startCase(get(relatedResources[0], 'product')),
          `${quotaItem.consumed} of ${quotaItem.allowed}`,
          { title: this.getCapacityIcon(quotaItem.consumed, quotaItem.allowed) },
        ]];
      });
    }

    // all rows may be filtered out if a user doesn't have any quota
    if (rows.length > 0) {
      content = (
        <>
          <StackItem className="content-header">
            Quota
          </StackItem>
          <StackItem className="table-container">
            <OSDSubscriptionTable rows={rows} />
          </StackItem>
        </>
      );
    } else {
      const data = {
        error: quotaCost.error,
        pending: quotaCost.pending,
        type: 'osd',
        empty: true,
      };
      content = <SubscriptionNotFulfilled data={data} refresh={this.refresh} />;
    }

    return (
      <Card>
        <CardTitle>OpenShift Dedicated</CardTitle>
        <CardBody>
          <Stack hasGutter>
            <StackItem>
              The summary of all subscriptions for OpenShift Dedicated purchased
              by your organization or granted by Red Hat.
            </StackItem>
            {content}
          </Stack>
        </CardBody>
      </Card>
    );
  }
}

OSDSubscriptionCard.defaultProps = {
  marketplaceQuotaFeature: false,
};

OSDSubscriptionCard.propTypes = {
  organizationID: PropTypes.string.isRequired,
  fetchQuotaCost: PropTypes.func.isRequired,
  quotaCost: PropTypes.object.isRequired,
  marketplaceQuotaFeature: PropTypes.bool,
};

export default OSDSubscriptionCard;
