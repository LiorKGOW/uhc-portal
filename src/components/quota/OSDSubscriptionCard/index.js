import { connect } from 'react-redux';

import { subscriptionsActions } from '../../../redux/actions/subscriptionsActions';
import OSDSubscriptionCard from './OSDSubscriptionCard';

import { MARKETPLACE_QUOTA_FEATURE } from '../../../redux/constants/featureConstants';

const mapDispatchToProps = {
  fetchQuotaCost: organizationID => subscriptionsActions.fetchQuotaCost(organizationID),
};

function mapStateToProps(state) {
  return {
    quotaCost: state.subscriptions.quotaCost,
    marketplaceQuotaFeature: state.features[MARKETPLACE_QUOTA_FEATURE],
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(OSDSubscriptionCard);
