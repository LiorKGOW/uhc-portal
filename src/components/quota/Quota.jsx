/*
Copyright (c) 2018 Red Hat, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { PageHeader, PageHeaderTitle } from '@redhat-cloud-services/frontend-components';
import { PageSection, Stack, StackItem } from '@patternfly/react-core';
import get from 'lodash/get';

import OSDSubscriptionCard from './OSDSubscriptionCard';
import SubscriptionNotFulfilled from './SubscriptionNotFulfilled';
import './Quota.scss';

class Subscriptions extends Component {
  componentDidMount() {
    document.title = 'Quota | Red Hat OpenShift Cluster Manager';
    this.refresh();
    if (get(window, 'insights.ocm')) {
      this.cleanupOcmListener = insights.ocm.on('APP_REFRESH', () => this.refresh());
    }
  }

  componentWillUnmount() {
    const { invalidateClusters } = this.props;
    invalidateClusters();

    if (this.cleanupOcmListener) {
      this.cleanupOcmListener();
    }
  }

  refresh = () => {
    const { fetchAccount } = this.props;
    fetchAccount();
  };

  render() {
    const { account, marketplace } = this.props;
    let content;
    let title = 'Dedicated (Annual)';
    if (marketplace) {
      title = 'Dedicated (On-Demand Limits)';
    }
    if (account.fulfilled && account.data.organization && account.data.organization.id) {
      const organizationID = account.data.organization.id;
      content = (
        <>
          <PageHeader>
            <PageHeaderTitle title={title} className="page-title" />
          </PageHeader>
          <PageSection className="ocm-p-subscriptions">
            <Stack hasGutter>
              <StackItem className="ocm-l-osd-subscription__section">
                <OSDSubscriptionCard organizationID={organizationID} marketplace={marketplace} />
              </StackItem>
            </Stack>
          </PageSection>
        </>
      );
    } else {
      const data = {
        error: account.error,
        pending: account.pending,
        type: 'account',
        internalErrorCode: account.internalErrorCode,
      };
      content = <SubscriptionNotFulfilled data={data} refresh={this.refresh} />;
    }

    return content;
  }
}

Subscriptions.propTypes = {
  fetchAccount: PropTypes.func.isRequired,
  account: PropTypes.object.isRequired,
  invalidateClusters: PropTypes.func.isRequired,
  marketplace: PropTypes.bool,
};

export default Subscriptions;
