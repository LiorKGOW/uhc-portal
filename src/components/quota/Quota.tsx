/*
Copyright (c) 2018 Red Hat, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import React from 'react';
import { PageHeader, PageHeaderTitle } from '@redhat-cloud-services/frontend-components/PageHeader';
import { PageSection, Stack, StackItem } from '@patternfly/react-core';

import OSDSubscriptionCard from './OSDSubscriptionCard';
import SubscriptionNotFulfilled from './SubscriptionNotFulfilled';
import type { State as SubscriptionState } from '../../redux/reducers/subscriptionsReducer';

import './Quota.scss';
import { AppPage } from '../App/AppPage';

type Props = {
  fetchAccount: () => void;
  account: SubscriptionState['account'];
  invalidateClusters: () => void;
  marketplace?: boolean;
};

const PAGE_TITLE = 'Quota | Red Hat OpenShift Cluster Manager';

const Quota = ({ invalidateClusters, fetchAccount, account, marketplace }: Props) => {
  // store in refs so that we can retrieve the latest value without re-creating the effect
  const fetchAccountRef = React.useRef(fetchAccount);
  fetchAccountRef.current = fetchAccount;
  const invalidateClustersRef = React.useRef(invalidateClusters);
  invalidateClustersRef.current = invalidateClusters;

  React.useEffect(() => {
    fetchAccount();
    const cleanupOcmListener = window.insights?.ocm?.on('APP_REFRESH', () =>
      fetchAccountRef.current(),
    );
    return () => {
      invalidateClustersRef.current();

      if (cleanupOcmListener) {
        cleanupOcmListener();
      }
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (account.fulfilled && account.data.organization && account.data.organization.id) {
    const title = marketplace ? 'Dedicated (On-Demand Limits)' : 'Dedicated (Annual)';
    const organizationID = account.data.organization.id;
    return (
      <AppPage title={PAGE_TITLE}>
        <PageHeader>
          <PageHeaderTitle title={title} className="page-title" />
        </PageHeader>
        <PageSection className="ocm-p-subscriptions">
          <Stack hasGutter>
            <StackItem className="ocm-l-osd-subscription__section">
              <OSDSubscriptionCard organizationID={organizationID} marketplace={marketplace} />
            </StackItem>
          </Stack>
        </PageSection>
      </AppPage>
    );
  }
  return (
    <AppPage title={PAGE_TITLE}>
      <SubscriptionNotFulfilled
        data={{
          error: account.error,
          pending: account.pending,
          type: 'account',
          internalErrorCode: account.error ? account.internalErrorCode : undefined,
        }}
        refresh={fetchAccount}
      />
    </AppPage>
  );
};

export default Quota;
