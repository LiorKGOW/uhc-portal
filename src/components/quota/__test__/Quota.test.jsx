import React from 'react';
import { shallow } from 'enzyme';
import { Button } from '@patternfly/react-core';

import * as Fixtures from './Quota.fixtures';
import Quota from '../Quota';
import OSDSubscriptionCard from '../OSDSubscriptionCard/OSDSubscriptionCard';
import OSDSubscriptionTable from '../OSDSubscriptionCard/OSDSubscriptionTable';
import SubscriptionNotFulfilled from '../SubscriptionNotFulfilled';

describe('<Quota />', () => {
  let wrapper;
  let refreshFn;
  describe('Quota', () => {
    beforeEach(() => {
      wrapper = shallow(<Quota {...Fixtures} />);
    });

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should call fetch method', () => {
      expect(Fixtures.fetchAccount).toBeCalled();
    });
    it('should have Header, OCP and OSD cards', () => {
      expect(wrapper.find('PageHeader').length).toEqual(1);
      expect(wrapper.find('Connect(OSDSubscriptionCard)').length).toEqual(1);
    });
  });

  describe('OSDSubscriptionCard', () => {
    beforeEach(() => {
      wrapper = shallow(<OSDSubscriptionCard {...Fixtures} />);
    });

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should call fetch method', () => {
      expect(Fixtures.fetchQuotaCost).toBeCalled();
    });
    it('should have OSDSubscriptionTable', () => {
      const tableComponent = wrapper.find('OSDSubscriptionTable');
      expect(tableComponent.length).toEqual(1);
      expect(tableComponent.props().rows.length).toEqual(Fixtures.expectedRowsForQuotaCost);
    });
  });

  describe('OSDSubscriptionTable', () => {
    beforeEach(() => {
      wrapper = shallow(<OSDSubscriptionTable {...Fixtures} />);
    });

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  // TODO: following tests should test OSDSubscriptionCard rather than
  //   directly calling SubscriptionNotFulfilled.
  describe('OSDSubscriptionCard Loading', () => {
    const quotaCost = { ...Fixtures.quotaCost, pending: true, type: 'osd' };
    beforeEach(() => {
      refreshFn = jest.fn();
      wrapper = shallow(<SubscriptionNotFulfilled data={quotaCost} refresh={refreshFn} />);
    });

    it('should render loading OSD quota summary', () => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('Spinner').length).toEqual(1);
    });
  });

  describe('OSDSubscriptionCard Empty', () => {
    const quotaCost = { ...Fixtures.emptyQuotaCost, empty: true, type: 'osd' };
    beforeEach(() => {
      refreshFn = jest.fn();
      wrapper = shallow(<SubscriptionNotFulfilled data={quotaCost} refresh={refreshFn} />);
    });

    it('should render empty OSD quota summary', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('OSDSubscriptionCard Error', () => {
    const quotaCost = { ...Fixtures.quotaCost, error: true, type: 'osd' };
    beforeEach(() => {
      refreshFn = jest.fn();
      wrapper = shallow(<SubscriptionNotFulfilled data={quotaCost} refresh={refreshFn} />);
    });

    it('should render error OSD quota summary', () => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find(Button).length).toEqual(1);
      wrapper.find(Button).simulate('click');
      expect(refreshFn).toBeCalled();
    });
  });
});
