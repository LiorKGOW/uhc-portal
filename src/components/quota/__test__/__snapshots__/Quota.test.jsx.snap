// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Quota /> OSDSubscriptionCard Empty should render empty OSD quota summary 1`] = `
<PageSection
  className="subscriptions-empty-state"
>
  <EmptyState>
    <Title
      headingLevel="h4"
      size="2xl"
    >
      You do not have any quota
    </Title>
    <EmptyStateBody>
      <p>
        <a
          href="https://cloud.redhat.com/products/dedicated/contact/"
          rel="noreferrer noopener"
          target="_blank"
        >
          Contact sales
        </a>
         
        to get started with OpenShift Dedicated.
      </p>
    </EmptyStateBody>
  </EmptyState>
</PageSection>
`;

exports[`<Quota /> OSDSubscriptionCard Error should render error OSD quota summary 1`] = `
<PageSection
  className="subscriptions-empty-state"
>
  <EmptyState>
    <Title
      headingLevel="h4"
      size="2xl"
    >
      Unable to retrieve quota information
    </Title>
    <EmptyStateBody>
      <p>
        An error has occured! Try again or contact support by including this error message:
      </p>
      <q />
      <p>
        Operation ID: N/A
      </p>
    </EmptyStateBody>
    <Button
      onClick={[MockFunction]}
    >
      Try again
    </Button>
  </EmptyState>
</PageSection>
`;

exports[`<Quota /> OSDSubscriptionCard Loading should render loading OSD quota summary 1`] = `
<PageSection>
  <Spinner
    centered={true}
  />
</PageSection>
`;

exports[`<Quota /> OSDSubscriptionCard should render 1`] = `
<Card>
  <CardTitle>
    OpenShift Dedicated
  </CardTitle>
  <CardBody>
    <Stack
      hasGutter={true}
    >
      <StackItem>
        The summary of all annual subscriptions for OpenShift Dedicated purchased by your organization or granted by Red Hat. For on-demand resources, see
         
        <Link
          to="/quota/resource-limits"
        >
          Dedicated (On-Demand Limits)
        </Link>
      </StackItem>
      <StackItem
        className="content-header"
      >
        Quota
      </StackItem>
      <StackItem
        className="table-container"
      >
        <OSDSubscriptionTable
          rows={
            Array [
              Array [
                "cluster",
                "gp.small",
                Object {
                  "title": "multizone",
                },
                "Standard",
                "OSD",
                "0 of 17",
                Object {
                  "title": <OutlinedCircleIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                },
              ],
              Array [
                "compute.node",
                "gp.small",
                Object {
                  "title": "N/A",
                },
                "Standard",
                "OSD",
                "4 of 27",
                Object {
                  "title": <ResourcesAlmostEmptyIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                },
              ],
              Array [
                "add-on",
                "addon-dba-operator",
                Object {
                  "title": "N/A",
                },
                "Any",
                "ANY",
                "0 of 15",
                Object {
                  "title": <OutlinedCircleIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="sm"
                  />,
                },
              ],
            ]
          }
        />
      </StackItem>
    </Stack>
  </CardBody>
</Card>
`;

exports[`<Quota /> OSDSubscriptionTable should render 1`] = `
<Fragment>
  <Table
    aria-label="Quota Table"
    borders={true}
    canCollapseAll={false}
    canSelectAll={true}
    canSortFavorites={true}
    cells={
      Array [
        "Resource type",
        "Resource name",
        "Availability",
        Object {
          "title": "Plan type",
          "transforms": Array [
            [Function],
          ],
        },
        "Cluster type",
        Object {
          "props": Object {
            "className": "quota-table-used",
          },
          "title": "Used",
        },
        "Capacity",
      ]
    }
    className=""
    collapseAllAriaLabel=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    isHeaderSelectDisabled={false}
    isNested={false}
    isStickyHeader={false}
    isTreeTable={false}
    ouiaSafe={true}
    role="grid"
    rowLabeledBy="simple-node"
    rows={Array []}
    selectVariant="checkbox"
    variant={null}
  >
    <TableHeader />
    <TableBody />
  </Table>
</Fragment>
`;

exports[`<Quota /> Quota should render 1`] = `
<Fragment>
  <PageHeader>
    <PageHeaderTitle
      className="page-title"
      title="Dedicated (Annual)"
    />
  </PageHeader>
  <PageSection
    className="ocm-p-subscriptions"
  >
    <Stack
      hasGutter={true}
    >
      <StackItem
        className="ocm-l-osd-subscription__section"
      >
        <Connect(OSDSubscriptionCard)
          organizationID="org-1"
        />
      </StackItem>
    </Stack>
  </PageSection>
</Fragment>
`;
